libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+void AP_RunCam::exit_2_key_osd_menu()
+{
+    // turn built-in OSD back on
+    enable_osd();
+
+    if (_video_recording) {
+        simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_START_RECORDING);
+        set_mode_change_timeout();
+    }
+    _state = State::READY;
+    _in_menu = 0;
+}
+
+// run the 5-key OSD simulation process
+void AP_RunCam::handle_5_key_simulation_process(Event ev) 
+{
+    switch (ev) {
+    case Event::BUTTON_RELEASE:
+        send_5_key_OSD_cable_simulation_event(ev);
+        _waiting_device_response = true;
+        return;
+        break;
+    case Event::ENTER_MENU:
+        if (_in_menu > 0) {
+            ev = Event::IN_MENU_RIGHT;
+        }
+        else {
+            // turn off built-in OSD so that the runcam OSD is visible
+            disable_osd();
+        }
+        break;
+
+    case Event::EXIT_MENU:
+        if (_in_menu > 0) {
+            // turn built-in OSD back on
+            enable_osd();
+        }
+        break;
+
+    default:
+        break;
+    }
+
+    if (ev != Event::NONE) {
+        send_5_key_OSD_cable_simulation_event(ev);
+        _is_button_pressed = true;
+        _waiting_device_response = true;
+    }
+}
+
+// handle a response
+void AP_RunCam::handle_5_key_simulation_response(const Response& response)
+{
+    if (response._result != ResponseStatus::SUCCESS) {
+        debug("response for command %d failed: %d\n", int(response._command), int(response._result));
+        simulation_OSD_cable_failed(response);
+        _waiting_device_response = false;
+        return;
+    }
+
+    debug("received response for command %d\n", int(response._command));
+    switch (response._command) {
+    case Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_RELEASE:
+        _is_button_pressed = false;
+        break;
+    case Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION:
+    {
+        // the high 4 bits is the operationID that we sent
+        // the low 4 bits is the result code
+        _is_button_pressed = true;
+        ConnectionOperation operationID = ConnectionOperation(response._param);
+        uint8_t errorCode = (response._recv_buf[1] & 0x0F);
+        switch (operationID) {
+        case ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_OPEN:
+            if (errorCode > 0) {
+                _in_menu = 1;
+            }
+            break;
+        case ConnectionOperation::RCDEVICE_PROTOCOL_5KEY_FUNCTION_CLOSE:
+            if (errorCode > 0) {
+                _in_menu = 0;
+            }
+            break;
+        }
+        break;
+    }
+    case Command::RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_PRESS:
+        _is_button_pressed = true;
+        break;
+    default:
+        break;
+    }
+
+    _waiting_device_response = false;
+}
+
+// process a response from the serial port
+void AP_RunCam::receive()
+{
+    if (!uart) {
+        return;
+    }
+
+    Response response;
+    while (process_waiting(response, AP_HAL::millis())) {
```

Comment:
timeouts appear to be few and far between here.  Long-runninbg loops are bad on flight controllers, so we like to be over-cautious on out loops.  Explicit max-duration or limiting the number of bytes processed are typical.

Reply:
I'm finding the logic here rather difficult to follow.  Could you outline what you're trying to achieve here with this queue?

Reply:
Removed so fixed